cardinal/ecs/component.go

```diff
@@ -225,17 +225,19 @@ func GetComponent[T component.INameable](w *World, id entity.ID) (comp *T, err e
 	}
 	value, err := w.StoreManager().GetComponentForEntity(c, id)
 	if err != nil {
-		return comp, err
+		return nil, err
 	}
-	comp, ok = value.(*T)
+	t, ok = value.(T)
 	if !ok {
-		return comp, fmt.Errorf("type assertion for component failed: %v to %v", value, c)
+		return nil, fmt.Errorf("type assertion for component failed: %v to %v", value, c)
 	}
+	comp = &t
+
 	return comp, nil
 }
 
 // Set sets component data to the entity.
-func SetComponent[T component.INameable](w *World, id entity.ID, component any) error {
+func SetComponent[T component.INameable](w *World, id entity.ID, component *T) error {
```

Comment:
The final `component *T` should just be `component T`. It should just match the exact type that the user initially passes in.

Reply:
Are you sure? This will lead to a copy. 

If I have: 

`type SomethingWithmanyFields {
    field1 int
    ...
    field99 int
}`

That leads to a full copy. More efficient to pass the pointer.

Reply:
it'll probably be more common to have a component struct with at most a handful of fields (1-5) - components with _many_ fields would go against ECS design patterns

Reply:
Also, there's nothing stopping a user from using a pointer as the T (which would result in copies by reference). IMO, it's confusing to ask for the user to supply a type and then de-reference it.

Reply:
> it'll probably be more common to have a component struct with at most a handful of fields (1-5) - components with _many_ fields would go against ECS design patterns

Yeah of course, but I'm just pointing out worst case scenario here to make it obvious the issue with the copy.

Reply:
> Also, there's nothing stopping a user from using a pointer as the T (which would result in copies by reference). IMO, it's confusing to ask for the user to supply a type and then de-reference it.

But if I don't specify the pointer type in the function parameter And the user doesn't use a pointer as the T, then a pass by COPY becomes mandatory.

Reply:
> We want to make it mandatory that the the only thing passed here is a pointer.

Go supports passing data around by value and passing data around by pointers. There are tradeoffs; passing by pointer isn't always the best option. We want to leave that choice to the user.

Reply:
Go supports defining what function is pass by value or pass by pointer. We define the function here and since we know exactly what this function does We know that a pass by value here results in an unnecessary copy. 

Yes the user should have the choice if they defined the function and they know that a pass by copy is necessary. The key difference here is we are the ones defining the function. The user should not be given a choice here since a pass by copy results in a functionally pointless copy. 

Why would a user need a pass by copy here?

Reply:
Maybe they want to be confident the passed in value won't be modified. Maybe the struct size ends up being smaller than a pointer. Maybe they've set up methods on their structs to not use pointers and they want to be consistent.

Again, with a signature of:
```
func SetComponent[T component.INameable](w *World, id entity.ID, component T) error {
```
The user is allowed to do:
```
SetComponent[*Foo](world, 1, &Foo{})
```

to get the benefits you're talking about.

Reply:
>Maybe the struct size ends up being smaller than a pointer.

The chances of this are too minuscule to be practical imo. More likely the user will be accidentally copying big structs everywhere. 

>Maybe they want to be confident the passed in value won't be modified.

I disagree. We don't modify that value during a set, they can look at the code. I don't think a user should be given an option to not pass a pointer because their nervous about something out code doesn't do. 

>to get the benefits you're talking about.

Users are unlikely to to user`SetComponent[*Foo](...)
`. They will most likely do: SetComponent[Foo](...) as that seems more natural. Then because of this they will always be copying the struct. 

@technicallyty what do you think?

Reply:
I don't think the user will be passing around big structs - the most common case will be small structs

Reply:
@technicallyty the big struct thing is is just an extreme case but this does effect less extreme cases. 

Here's the thing. If Set is called a bunch of times even a component with say 10 fields becomes 10x slower. This magnifies the more set is called. So if Set is called 30 times, or 20 times it becomes 200 copies or 300 copies vs. 30 or 20. There is an O(N) slow down here vs. O(1).

Reply:
sounds like pointer is faster. might be nice to have benchmarks to affirm later on

Reply:
@jerargus I'm gonna go with the majority votes for now. We can always change this back later.